// env-runner/dist/_chunks/virtual-loader.mjs
async function resolveVirtualModules(virtual) {
	const entries = await Promise.all(Object.entries(virtual).map(async ([key, value]) => [key, typeof value === "function" ? await value() : value]));
	return Object.fromEntries(entries);
}
const VIRTUAL_SCHEME = "virtual:";
function createVirtualHooks(virtual) {
	const resolve = (specifier, context, nextResolve) => {
		if (Object.hasOwn(virtual, _stripQuery(specifier))) return {
			url: VIRTUAL_SCHEME + encodeURIComponent(specifier),
			shortCircuit: true
		};
		return nextResolve(specifier, context);
	};
	const load = (url, context, nextLoad) => {
		if (url.startsWith(VIRTUAL_SCHEME)) {
			const key = _stripQuery(decodeURIComponent(url.slice(8)));
			if (Object.hasOwn(virtual, key)) return {
				format: virtualModuleFormat(key),
				source: virtual[key],
				shortCircuit: true
			};
		}
		return nextLoad(url, context);
	};
	return {
		resolve,
		load
	};
}
function virtualModuleFormat(specifier) {
	if (specifier.endsWith(".json")) return "json";
	if (specifier.endsWith(".ts") || specifier.endsWith(".mts")) return "module-typescript";
	return "module";
}
function _stripQuery(specifier) {
	const qIndex = specifier.indexOf("?");
	return qIndex === -1 ? specifier : specifier.slice(0, qIndex);
}
export { createVirtualHooks, resolveVirtualModules, virtualModuleFormat };
