// env-runner/dist/_chunks/worker-utils.mjs
import { createVirtualHooks, virtualModuleFormat } from "./virtual-loader.mjs";
import { existsSync, readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";
import { isAbsolute } from "node:path";
async function registerVirtualModules(virtual) {
	if (!virtual || Object.keys(virtual).length === 0) return _noop;
	const { registerHooks, stripTypeScriptTypes } = await import("node:module");
	if (typeof registerHooks === "function") {
		if ("Deno" in globalThis) virtual = _transformForDeno(virtual, stripTypeScriptTypes);
		const hooks = registerHooks(createVirtualHooks(virtual));
		return _once(() => hooks.deregister());
	}
	if (typeof globalThis.Bun?.plugin === "function") {
		_bunVirtual = virtual;
		_registerBunModules(Object.keys(virtual));
		return _once(() => {
			if (_bunVirtual === virtual) _bunVirtual = void 0;
		});
	}
	console.warn("[env-runner] virtual modules require `module.registerHooks` (Node.js >= 22.15 / Deno >= 2.x) or `Bun.plugin`; skipping registration.");
	return _noop;
}
function refreshVirtualModule(specifier) {
	if (_bunVirtual?.[specifier] === void 0) return false;
	_registerBunModules([specifier]);
	return true;
}
let _bunVirtual;
function _registerBunModules(specifiers) {
	globalThis.Bun.plugin({
		name: "env-runner-virtual",
		setup(build) {
			for (const specifier of specifiers) build.module(specifier, () => {
				const source = _bunVirtual?.[specifier];
				if (source === void 0) throw new Error(`Cannot find virtual module "${specifier}" (unregistered)`);
				const format = virtualModuleFormat(specifier);
				if (format === "json") return {
					exports: { default: JSON.parse(source) },
					loader: "object"
				};
				return {
					contents: source,
					loader: format === "module-typescript" ? "ts" : "js"
				};
			});
		}
	});
}
function _transformForDeno(virtual, stripTypeScriptTypes) {
	const out = {};
	for (const [specifier, source] of Object.entries(virtual)) {
		const format = virtualModuleFormat(specifier);
		if (format === "module-typescript") {
			if (typeof stripTypeScriptTypes !== "function") throw new TypeError(`[env-runner] virtual TypeScript module "${specifier}" requires \`module.stripTypeScriptTypes\` (custom load hooks bypass Deno's native type stripping); upgrade Deno or provide a pre-transpiled JavaScript source instead.`);
			out[specifier] = stripTypeScriptTypes(source);
		} else if (format === "json") out[specifier] = `export default JSON.parse(${JSON.stringify(source)});`;
		else out[specifier] = source;
	}
	return out;
}
const _noop = () => {};
function _once(fn) {
	let done = false;
	return () => {
		if (!done) {
			done = true;
			fn();
		}
	};
}
function isVirtualSpecifier(specifier, virtual) {
	return Boolean(specifier && virtual && Object.hasOwn(virtual, specifier));
}
async function resolveEntry(entryPath, virtual) {
	const mod = await (virtual ? import(entryPath) : import(_toImportPath(entryPath)));
	const entry = mod.default || mod;
	if (typeof entry.fetch !== "function") throw new Error(`[env-runner] Entry module "${entryPath}" must export a \`fetch\` handler (export default { fetch(req) { ... } }).`);
	return entry;
}
function parseServerAddress(server) {
	const url = new URL(server.url);
	return {
		host: url.hostname,
		port: Number(url.port)
	};
}
async function reloadEntryModule(entryPath, currentEntry, sendMessage, virtual) {
	await currentEntry.ipc?.onClose?.();
	const newEntry = await _importFresh(entryPath, virtual);
	await newEntry.ipc?.onOpen?.({ sendMessage });
	return newEntry;
}
function _toImportPath(entryPath) {
	const qIndex = entryPath.indexOf("?");
	const filePath = qIndex === -1 ? entryPath : entryPath.slice(0, qIndex);
	const query = qIndex === -1 ? "" : entryPath.slice(qIndex);
	if (isAbsolute(filePath)) return pathToFileURL(filePath).href + query;
	return entryPath;
}
let _reloadCounter = 0;
async function _importFresh(entryPath, virtual) {
	const qIndex = entryPath.indexOf("?");
	const filePath = qIndex === -1 ? entryPath : entryPath.slice(0, qIndex);
	let mod;
	if (!virtual && existsSync(filePath)) {
		const code = readFileSync(filePath, "utf8");
		mod = await import("data:text/javascript;base64," + Buffer.from(code).toString("base64"));
	} else if (virtual && refreshVirtualModule(filePath)) mod = await import(filePath);
	else mod = await import(entryPath + (qIndex === -1 ? "?" : "&") + "__envRunnerReload=" + _reloadCounter++);
	const entry = mod.default || mod;
	if (typeof entry.fetch !== "function") throw new Error(`[env-runner] Entry module "${entryPath}" must export a \`fetch\` handler (export default { fetch(req) { ... } }).`);
	return entry;
}
export { isVirtualSpecifier, parseServerAddress, registerVirtualModules, reloadEntryModule, resolveEntry };
